kani-driver/src/list/output.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// This module handles outputting the result for the list subcommand
+
+use std::{
+    cmp::max,
+    collections::{BTreeMap, BTreeSet},
+    fs::File,
+    io::BufWriter,
+};
+
+use crate::{list::Totals, version::KANI_VERSION};
+use anyhow::Result;
+use colour::print_ln_bold;
+use kani_metadata::ContractedFunction;
+use serde_json::json;
+
+// Represents the version of our JSON file format.
+// Increment this version (according to semantic versioning rules) whenever the JSON output format changes.
+const FILE_VERSION: &str = "0.1";
+const JSON_FILENAME: &str = "kani-list.json";
+
+/// Construct the table of contracts information.
+fn construct_contracts_table(
+    contracted_functions: BTreeSet<ContractedFunction>,
+    totals: Totals,
+) -> Vec<String> {
+    const NO_HARNESSES_MSG: &str = "NONE";
+
+    // Since the harnesses will be separated by newlines, the column length is equal to the length of the longest harness
+    fn column_len(harnesses: &[String]) -> usize {
+        harnesses.iter().map(|s| s.len()).max().unwrap_or(NO_HARNESSES_MSG.len())
+    }
+
+    // Contracts table headers
+    const FUNCTION_HEADER: &str = "Function";
+    const CONTRACT_HARNESSES_HEADER: &str = "Contract Harnesses (#[kani::proof_for_contract])";
+
+    // Contracts table totals row
+    const TOTALS_HEADER: &str = "Total";
+    let functions_total = totals.contracted_functions.to_string();
+    let harnesses_total = totals.contract_harnesses.to_string();
+
+    let mut table_rows: Vec<String> = vec![];
+    let mut max_function_fmt_width = max(FUNCTION_HEADER.len(), functions_total.len());
+    let mut max_contract_harnesses_fmt_width =
+        max(CONTRACT_HARNESSES_HEADER.len(), harnesses_total.len());
+
+    let mut data_rows: Vec<(String, Vec<String>)> = vec![];
+
+    for cf in contracted_functions {
+        max_function_fmt_width = max(max_function_fmt_width, cf.function.len());
+        max_contract_harnesses_fmt_width =
+            max(max_contract_harnesses_fmt_width, column_len(&cf.harnesses));
+
+        data_rows.push((cf.function, cf.harnesses));
+    }
```

Comment:
I'm not quite sure what you mean--I could write syntactically different code to achieve this same goal, but in any case it would still require iterating through each `ContractedFunction` to find the maximum widths.
I agree that this code is a bit long. I took inspiration for this format from [Adrian's `kani-cov` work](https://github.com/model-checking/kani/pull/3121). I can see a future where we have a new crate in `kani-driver` that just prints Markdown tables given inputs, which would abstract away some of this detail.